src/pipeformer/internal/structures.py

```diff
@@ -0,0 +1,298 @@
+
+    provider: str = attr.ib(validator=instance_of(str))
+    inputs: Set[str] = attr.ib(
+        default=attr.Factory(set), validator=optional(deep_iterable(member_validator=instance_of(str)))
+    )
+    outputs: Set[str] = attr.ib(
+        default=attr.Factory(set), validator=optional(deep_iterable(member_validator=instance_of(str)))
+    )
+    configuration: Dict[str, str] = attr.ib(default=attr.Factory(dict), validator=optional(_STRING_STRING_MAP))
+    image: Optional[str] = attr.ib(default=None, validator=optional(instance_of(str)))
+    environment_type: Optional[str] = attr.ib(default=None, validator=optional(instance_of(str)))
+    buildspec: Optional[str] = attr.ib(default=None, validator=optional(instance_of(str)))
+    compute_type: str = attr.ib(default="BUILD_GENERAL1_SMALL", validator=optional(instance_of(str)))
+    env: Dict[str, str] = attr.ib(default=attr.Factory(dict), validator=optional(_STRING_STRING_MAP))
+    run_order: int = attr.ib(default=1, validator=optional(instance_of(int)))
+
+    @run_order.validator
+    def _check_run_order(self, attribute, value):  # pylint: disable=unused-argument,no-self-use
+        """Verify that ``run_order`` value is valid."""
+        if value < 1:
+            raise ValueError("PipelineAction run_order value must be >= 1")
+
+    @image.validator
+    def _check_image(self, attribute, value):  # pylint: disable=unused-argument
+        """Verify that ``image`` is set if provider type ``CodeBuild`` is used."""
+        if self.provider == "CodeBuild" and value is None:
+            raise ValueError('image must be defined for actions of type "CodeBuild"')
+
+    @buildspec.validator
+    def _check_buildspec(self, attribute, value):  # pylint: disable=unused-argument
+        """Verify that ``buildspec`` is set if provider type ``CodeBuild`` is used."""
+        if self.provider == "CodeBuild" and value is None:
+            raise ValueError('buildspec must be defined for actions of type "CodeBuild"')
+
+    def __attrs_post_init__(self):
+        """Set default values for ``environment_type``."""
+        if self.provider == "CodeBuild" and self.environment_type is None:
+            if "windows" in self.image.lower():
+                self.environment_type = "WINDOWS_CONTAINER"
+            else:
+                self.environment_type = "LINUX_CONTAINER"
+
+
+@attr.s
+class PipelineStage(_ConfigStructure):
+    """CodePipeline stage definition.
+
+    :param name: Stage name
+    :param actions: Actions to be taken in stage
+    """
+
+    name: str = attr.ib(validator=instance_of(str))
+    actions: Iterable[PipelineAction] = attr.ib(validator=deep_iterable(member_validator=instance_of(PipelineAction)))
+
+
+@attr.s
+class Config(_ConfigStructure):
+    """PipeFormer project configuration.
+
+    :param name: Project name
+    :param description: Project description
+    :param custom_cmk: Should a custom CMK be generated? (reserved for later use: must always be ``True``)
```

Comment:
Reading here and elsewhere in the code, this looks like it could both be a "custom" CMK (i.e. not something generated by Pipeformer?) or a "**custom**er-managed CMK", which makes it a little harder to grok what's intended. Could you clarify this and maybe rename if needed to disambiguate?

Reply:
Changed to `generate_cmk`, since this is meant to communicate the intent of "generate a unique CMK for this project".